pkg/operator/controller/dnsnameresolver/controller.go

```diff
@@ -0,0 +1,217 @@
+	// Create a new cache for tracking the DNSNameResolver resources in
+	// the DNSNameResolverNamespace.
+	dnsNameResolverCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+		DefaultNamespaces: map[string]cache.Config{
+			config.DNSNameResolverNamespace: {},
+		},
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create a new cache to track the EndpointSlices corresponding to the
+	// CoreDNS pods.
+	corednsEndpointsSliceCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+		DefaultNamespaces: map[string]cache.Config{
+			config.OperandNamespace: {},
+		},
+		DefaultLabelSelector: labels.SelectorFromSet(labels.Set{
+			discoveryv1.LabelServiceName: config.ServiceName,
+		}),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Initialize the reconciler.
+	reconciler := &reconciler{
+		config:               config,
+		dnsNameResolverCache: dnsNameResolverCache,
+		client:               mgr.GetClient(),
+		res:                  NewResolver(corednsEndpointsSliceCache, config.ServiceName),
+	}
+
+	// Create an unmanaged controller using the reconciler.
+	c, err := controller.NewUnmanaged(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Watch for the DNSNameResolver resources using dnsNameResolverCache.
+	if err := c.Watch(source.Kind(dnsNameResolverCache, &networkv1alpha1.DNSNameResolver{}), &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, nil, err
+	}
+
+	// Watch for the CoreDNS pod EndpointSlices using corednsEndpointsSliceCache.
+	if err := c.Watch(source.Kind(corednsEndpointsSliceCache, &discoveryv1.EndpointSlice{}), handler.EnqueueRequestsFromMapFunc(skipReconcileForEndpointSlice)); err != nil {
+		return nil, nil, err
+	}
+
+	return c, []cache.Cache{dnsNameResolverCache, corednsEndpointsSliceCache}, nil
+}
+
+// skipReconcileForEndpointSlice just skips the reconcile if any event related to the CoreDNS pod
+// endpointslices is received.
+func skipReconcileForEndpointSlice(context.Context, client.Object) []reconcile.Request {
+	return nil
+}
+
+type Config struct {
+	OperandNamespace         string
+	ServiceName              string
+	DNSNameResolverNamespace string
+}
+
+// reconciler handles the actual DNSNameResolver reconciliation logic in response to events.
+type reconciler struct {
+	config Config
+
+	dnsNameResolverCache cache.Cache
+	client               client.Client
+	res                  *Resolver
+	startResolver        sync.Once
+}
+
+// Reconcile expects request to refer to an DNSNameResolver resource, and will do all the work to
+// keep the status of the resource updated.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// Ensure that the resolver is started before doing anything else.
+	r.startResolver.Do(func() {
```

Comment:
It's not clear to me why this isn't done outside the reconcile step.

Reply:
The call to the `NewUnmanaged` function to create the unmanaged dnsnameresolver controller will always be executed even if the DNSNameResolver feature is not enabled. However, Reconcile will be called when the feature is enabled and the controller is started by the dnsnameresolvercrd controller. As, the resolver is needed to be started only when the feature is enabled, it is added to Reconcile.

Reply:
> The call to the NewUnmanaged function to create the unmanaged dnsnameresolver controller will always be executed even if the DNSNameResolver feature is not enabled.

I'm not convinced.  I'm going to ask @Miciah to take a quick look at this.

Reply:
We have similar logic in the ingress operator's canary controller: https://github.com/openshift/cluster-ingress-operator/blob/d31546e05cc00021e5fbd441ba6dc6739c1bf274/pkg/operator/controller/canary/controller.go#L191-L193

It's an unusual but established pattern for time-based reconciliation logic (for example, a periodic polling loop) as opposed to interrupt-based logic (for example, create/update/delete events):  

1. The controller watches some resource that should trigger the polling loop.  
2. The reconciliation loop ensures that resources that the polling loop requires are present.  
3. The reconciliation loop then starts the polling loop, using `sync.Once` and a goroutine.
4. The polling loop then can perform its work asynchronously from the reconciler.  

If the polling loop doesn't fit into controller-runtime's event model, then this approach seems reasonable.  However, I wonder how difficult it would be to use a more common pattern and avoid the explicit polling loop.  

@arkadeepsen, did you consider alternatives, such as using using `RequeueAfter` in the [`reconcile.Result`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.2/pkg/reconcile#Result), based on the next-to-expire TTL from the `DNSNameResolver` that you are reconciling?  Or another option, would it be any simpler to use a [channel source](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.2/pkg/source#Channel) to trigger reconciliations?  

If what you have is the simplest solution, then I don't see why we cannot use it, but if the same result can be achieved using more common controller-runtime patterns without causing significant inefficiency or complexity, then using more common patterns would most likely make for more readable, maintainable code.

Reply:
> @arkadeepsen, did you consider alternatives, such as using using `RequeueAfter` in the [`reconcile.Result`](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.2/pkg/reconcile#Result), based on the next-to-expire TTL from the `DNSNameResolver` that you are reconciling?

The `DNSNameResovler` CRs can be created for both regular and wildcard DNS names. Now a regular DNS name being looked up can match both the `DNSNameResolver` CRs and get added to the status of both the CRs. If we use `RequeueAfter` for each DNS name present in the status of the `DNSNameResolver` CRs, the refresh for a particular DNS name might get triggered more than once because of this reason.

It is easier to track the DNS names and the TTLs of the corresponding IP addresses in a separate periodic polling loop. Whenever the TTL of an IP address expires then the corresponding DNS name is looked up.

> Or another option, would it be any simpler to use a [channel source](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.17.2/pkg/source#Channel) to trigger reconciliations?

Actually a channel is being used but the other way round. Whenever a create/update event occurs, each DNS name present in the status of the `DNSNameResolver` object is added to the `Resolver`'s list of DNS names if not already added. This also needs to trigger a recalculation of the next expiring DNS name and the remaining time till the TTL expiration which is done via the channel.

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2